# The App Store and Google Play Ad Revenue Breakdown

We are looking to find the most profitable markets for mobile applications in the App Store and Google play markets. Look at various data sets, we can take a data-driven approach to developing new applications with high profit returns.

The goal of this project is to breakdown the markets to find the best approach to building new applications. As a company that only builds free apps, ads play a huge part in the decision-making process, as it is the source of income. Analyzing user bases and other various indicators will help influence the direction of development and marketing.

## Datasets

For our analysis, we are using two datasets:

[A Google Play dataset](https://www.kaggle.com/lava18/google-play-store-apps/home) featuring 10,000 apps from the Google Play Store, collected in August 2018.

[An App Store dataset](https://www.kaggle.com/ramamet4/app-store-apple-data-set-10k-apps/home) featuring 7,000 app form the App Store, collected in July 2017.



In [1]:
from csv import reader

# Google Play dataset
opened_file = open('googleplaystore.csv')
read_file = reader(opened_file)
google_data = list(read_file)
android_header = google_data[0]
android = google_data[1:]

# App Store dataset
opened_file = open('AppleStore.csv')
read_file = reader(opened_file)
apple_data = list(read_file)
ios_header = apple_data[0]
ios = apple_data[1:]

To start, we gather the datasets from both the Google Play Store and the App Store. To make things easier to reference, we separate the header from the rest of the data.

Let's look at the Play Store dataset.

In [2]:
def explore_data(dataset, first = 0, last = 5):
    data_section = dataset[first:last]
    for row in data_section:
        print(row)
        print('\n')
    print('Number of rows: ', len(dataset))
    print('Number of columns: ', len(dataset[0]))

print(android_header)
print('\n')
explore_data(android)

['App', 'Category', 'Rating', 'Reviews', 'Size', 'Installs', 'Type', 'Price', 'Content Rating', 'Genres', 'Last Updated', 'Current Ver', 'Android Ver']


['Photo Editor & Candy Camera & Grid & ScrapBook', 'ART_AND_DESIGN', '4.1', '159', '19M', '10,000+', 'Free', '0', 'Everyone', 'Art & Design', 'January 7, 2018', '1.0.0', '4.0.3 and up']


['Coloring book moana', 'ART_AND_DESIGN', '3.9', '967', '14M', '500,000+', 'Free', '0', 'Everyone', 'Art & Design;Pretend Play', 'January 15, 2018', '2.0.0', '4.0.3 and up']


['U Launcher Lite – FREE Live Cool Themes, Hide Apps', 'ART_AND_DESIGN', '4.7', '87510', '8.7M', '5,000,000+', 'Free', '0', 'Everyone', 'Art & Design', 'August 1, 2018', '1.2.4', '4.0.3 and up']


['Sketch - Draw & Paint', 'ART_AND_DESIGN', '4.5', '215644', '25M', '50,000,000+', 'Free', '0', 'Teen', 'Art & Design', 'June 8, 2018', 'Varies with device', '4.2 and up']


['Pixel Draw - Number Art Coloring Book', 'ART_AND_DESIGN', '4.3', '967', '2.8M', '100,000+', 'Free', '0', 'Eve

As you can see, there are 13 columns we can use for the dataset as well as 10841 rows of data. Of these columns, some of the most useful will be `'App'`, `'Category'`, `'Reviews'`, `'Installs'`, `'Type'`, `'Price'`, and `'Genres'`.

Next, we will look at the App Store data set.

In [3]:
print(ios_header)
print('\n')
explore_data(ios)

['id', 'track_name', 'size_bytes', 'currency', 'price', 'rating_count_tot', 'rating_count_ver', 'user_rating', 'user_rating_ver', 'ver', 'cont_rating', 'prime_genre', 'sup_devices.num', 'ipadSc_urls.num', 'lang.num', 'vpp_lic']


['284882215', 'Facebook', '389879808', 'USD', '0.0', '2974676', '212', '3.5', '3.5', '95.0', '4+', 'Social Networking', '37', '1', '29', '1']


['389801252', 'Instagram', '113954816', 'USD', '0.0', '2161558', '1289', '4.5', '4.0', '10.23', '12+', 'Photo & Video', '37', '0', '29', '1']


['529479190', 'Clash of Clans', '116476928', 'USD', '0.0', '2130805', '579', '4.5', '4.5', '9.24.12', '9+', 'Games', '38', '5', '18', '1']


['420009108', 'Temple Run', '65921024', 'USD', '0.0', '1724546', '3842', '4.5', '4.0', '1.6.2', '9+', 'Games', '40', '5', '1', '1']


['284035177', 'Pandora - Music & Radio', '130242560', 'USD', '0.0', '1126879', '3594', '4.0', '4.5', '8.4.1', '12+', 'Music', '37', '4', '1', '1']


Number of rows:  7197
Number of columns:  16


The App Store has 16 columns to use as well as 7197 rows of data to use. Some of the colummns we will use are `'track_name'`, `'currency'`, `'price'`, `'rating_count_tot'`, `'rating_count_ver'`, and `'prime_genre'`.

These columns are not self-explanitary, so to see what these columns do specifically, check out the [documentation](https://www.kaggle.com/ramamet4/app-store-apple-data-set-10k-apps/home).

## Data Cleaning

Before we get to far, we want to clean up the data before running. First, there is a row (#10472) in the Google Play Store that is incorrectly formatted, as referenced in the [discussion here](https://www.kaggle.com/lava18/google-play-store-apps/discussion/66015). Let's look at what is going on.

In [4]:
print(android[10472]) # incorrect row
print('\n')
print(android_header) # header
print('\n')
print(android[0]) # correct row

['Life Made WI-Fi Touchscreen Photo Frame', '1.9', '19', '3.0M', '1,000+', 'Free', '0', 'Everyone', '', 'February 11, 2018', '1.0.19', '4.0 and up']


['App', 'Category', 'Rating', 'Reviews', 'Size', 'Installs', 'Type', 'Price', 'Content Rating', 'Genres', 'Last Updated', 'Current Ver', 'Android Ver']


['Photo Editor & Candy Camera & Grid & ScrapBook', 'ART_AND_DESIGN', '4.1', '159', '19M', '10,000+', 'Free', '0', 'Everyone', 'Art & Design', 'January 7, 2018', '1.0.0', '4.0.3 and up']


We can see that the second column, Category, does not have an entry in row 10472. We will have to remove from our dataset.

In [5]:
print('Android columns before removal- ', len(android))
del android[10472] # removes from dataset
print('Android columns after removal- ', len(android))

Android columns before removal-  10841
Android columns after removal-  10840


## Duplicate Data

In the Google Play dataset, there are many instances of duplicate data, which needs to be handled.

In [6]:
for app in android:
    name = app[0]
    if name == 'Instagram':
        print(app)

['Instagram', 'SOCIAL', '4.5', '66577313', 'Varies with device', '1,000,000,000+', 'Free', '0', 'Teen', 'Social', 'July 31, 2018', 'Varies with device', 'Varies with device']
['Instagram', 'SOCIAL', '4.5', '66577446', 'Varies with device', '1,000,000,000+', 'Free', '0', 'Teen', 'Social', 'July 31, 2018', 'Varies with device', 'Varies with device']
['Instagram', 'SOCIAL', '4.5', '66577313', 'Varies with device', '1,000,000,000+', 'Free', '0', 'Teen', 'Social', 'July 31, 2018', 'Varies with device', 'Varies with device']
['Instagram', 'SOCIAL', '4.5', '66509917', 'Varies with device', '1,000,000,000+', 'Free', '0', 'Teen', 'Social', 'July 31, 2018', 'Varies with device', 'Varies with device']


As you can see, there are four entries for Instagram. There are many more instances of this throughout the dataset.

In [7]:
duplicate_apps = []
unique_apps = []

for app in android:
    name = app[0]
    if name in unique_apps:
        duplicate_apps.append(name)
    else:
        unique_apps.append(name)
        
print('Number of duplicate apps- ', len(duplicate_apps))
print('\n')
print('Some examples- ', duplicate_apps[:15])

Number of duplicate apps-  1181


Some examples-  ['Quick PDF Scanner + OCR FREE', 'Box', 'Google My Business', 'ZOOM Cloud Meetings', 'join.me - Simple Meetings', 'Box', 'Zenefits', 'Google Ads', 'Google My Business', 'Slack', 'FreshBooks Classic', 'Insightly CRM', 'QuickBooks Accounting: Invoicing & Expenses', 'HipChat - Chat Built for Teams', 'Xero Accounting Software']


As we can see, there is 1181 duplicate apps in our dataset. We will have to figure out a way to remove these from our dataset. 

If we look at the Instagram data above, we can see that the only difference between each row is the number of reviews. This will tell us that the data for each row was taken at different times. Instead of deleting rows at random, we can look at the highest humber of ratings to help guide us towards the most accurate row to use.

We will want to create a new dictionary where each entry is has a unique app name which has the highest amount of reviews, then use that dictionary as our new dataset.

First, we need to figure out how many rows we should see once we get rid of the duplicates.

In [8]:
print('Expected length- ', len(android) - len(duplicate_apps))

Expected length-  9659


We are expecting to see 9659 rows in our new dataset.

Now, we will make a new dictionary for each on the unique apps with the maximum amount of reviews.

In [9]:
reviews_max = {}

for app in android:
    name = app[0]
    n_reviews = float(app[3])
    
    if name in reviews_max and reviews_max[name] < n_reviews:
        reviews_max[name] = n_reviews
        
    elif name not in reviews_max:
        reviews_max[name] = n_reviews
        
print(len(reviews_max))

9659


As you can see, the length of this new dictionary is the same length as we expected from our previous calculations. We went through the android dataset and stored the name and the review amounts. If the name was already in the dictionary and the reviews that were there are less than the current row, we want to add the new value to the dictionary. Also, we want to add any name that is not yet added to the dictionary as well. So when a name is not in the dictionary, we will just simply add it. If it is already there, we want to test to see if the current value for the row is larger than the previous result. If it is, override the current selection.



In [10]:
android_clean = []
already_added = []

for app in android:
    name = app[0]
    n_reviews=float(app[3])
    
    if (reviews_max[name] == n_reviews) and (name not in already_added):
        android_clean.append(app)
        already_added.append(name)
        
print('Amount in android_clean- ', len(android_clean))

Amount in android_clean-  9659


To clean this database, we are going to create some containers, one for the clean database and one to compare the names to ones that are already added to the database.

We are then going to take the data in android and check the names and reviews. If the current review number is the same as what we have in the reviews_max dictionary, we know that it is the largest number of reviews. We also want to check to make sure that the name is not already in the already_added list. Once the criteria is met, we append the full row to the android_clean and the name to the already_added list.

We can run the explore data function to give us a quick look at this new dataset.

In [11]:
print(android_header)
print('\n')
explore_data(android_clean)

['App', 'Category', 'Rating', 'Reviews', 'Size', 'Installs', 'Type', 'Price', 'Content Rating', 'Genres', 'Last Updated', 'Current Ver', 'Android Ver']


['Photo Editor & Candy Camera & Grid & ScrapBook', 'ART_AND_DESIGN', '4.1', '159', '19M', '10,000+', 'Free', '0', 'Everyone', 'Art & Design', 'January 7, 2018', '1.0.0', '4.0.3 and up']


['U Launcher Lite – FREE Live Cool Themes, Hide Apps', 'ART_AND_DESIGN', '4.7', '87510', '8.7M', '5,000,000+', 'Free', '0', 'Everyone', 'Art & Design', 'August 1, 2018', '1.2.4', '4.0.3 and up']


['Sketch - Draw & Paint', 'ART_AND_DESIGN', '4.5', '215644', '25M', '50,000,000+', 'Free', '0', 'Teen', 'Art & Design', 'June 8, 2018', 'Varies with device', '4.2 and up']


['Pixel Draw - Number Art Coloring Book', 'ART_AND_DESIGN', '4.3', '967', '2.8M', '100,000+', 'Free', '0', 'Everyone', 'Art & Design;Creativity', 'June 20, 2018', '1.1', '4.4 and up']


['Paper flowers instructions', 'ART_AND_DESIGN', '4.4', '167', '5.6M', '50,000+', 'Free', '0', 'Every

As we can see, we have the same number of rows as we had before.

## Getting rid of non-English apps

As we are a company that deals in the English speaking market, we will need to remove apps that do not fit into our demographic. Let's look at a couple of examples of these non-English apps.

In [12]:
print(ios[813][1])
print(ios[6731][1])
print(android_clean[4412][0])
print(android_clean[7940][0])

爱奇艺PPS -《欢乐颂2》电视剧热播
【脱出ゲーム】絶対に最後までプレイしないで 〜謎解き＆ブロックパズル〜
中国語 AQリスニング
لعبة تقدر تربح DZ


As you can see, these apps would not be useful for our purposes.

To get rid of these apps, we will look for characters that are not in English. To do this, we will use the [ASCII](https://en.wikipedia.org/wiki/ASCII) system.

In [13]:
print(ord('奇'))
print(ord('a'))

22855
97


In the ASCII system, the numbers 0 to 127 are highly used English characters. Everything else is reserved for other things. We can then assume that anything without an English character is not what we want to include in out dataset.

In [14]:
def english_check(string):
    for character in string:
        if ord(character) > 127:
            return False
        
    return True

print(ios[0][1]," - ", english_check(ios[0][1]))
print(ios[813][1]," - ", english_check(ios[813][1]))
print('Instachat 😜 - ',english_check('Instachat 😜'))
print('Docs To Go™ Free Office Suite - ', english_check('Docs To Go™ Free Office Suite'))

Facebook  -  True
爱奇艺PPS -《欢乐颂2》电视剧热播  -  False
Instachat 😜 -  False
Docs To Go™ Free Office Suite -  False


Let's look at what is happening. We created a function called english_check to help break down each character in a title of an app. We pass in a string (in this case the name of the app) and go through each character and check if it is within the ASCII limit of 0 to 127 for the English language. If it is over that limit, it returns false. Otherwise, it returns true once it scans the entire string. Facebook returns true, while the other returns false. However, our other examples will also fall outside the ASCII range, even though it really should fall within. We would end up losing some data that we would actually want. Let's revise the function to help combat the issue.

In [15]:
def english_check(string):
    bad_character = 0
    
    for character in string:
        if ord(character) > 127:
            bad_character += 1
        
    if bad_character > 3:
        return False
    else:
        return True

print(ios[0][1]," - ", english_check(ios[0][1]))
print(ios[813][1]," - ", english_check(ios[813][1]))
print('Instachat 😜 - ', english_check('Instachat 😜'))
print('Docs To Go™ Free Office Suite - ', english_check('Docs To Go™ Free Office Suite'))

Facebook  -  True
爱奇艺PPS -《欢乐颂2》电视剧热播  -  False
Instachat 😜 -  True
Docs To Go™ Free Office Suite -  True


We modified the function to essentially add up the total of bad characters throughout the title. If there are more than 3, it is considered a bad string. There might still be some apps that fall through the cracks, but it will be close enough for our use.

Now let's cut out those non-English apps.

In [16]:
android_english = []
ios_english = []

for app in android_clean:
    name = app[0]
    checker = english_check(name)
    
    if checker:
        android_english.append(app)
        
for app in ios:
    name = app[1]
    checker = english_check(name)
    
    if checker:
        ios_english.append(app)
        
explore_data(android_english)
print('\n')
explore_data(ios_english)

['Photo Editor & Candy Camera & Grid & ScrapBook', 'ART_AND_DESIGN', '4.1', '159', '19M', '10,000+', 'Free', '0', 'Everyone', 'Art & Design', 'January 7, 2018', '1.0.0', '4.0.3 and up']


['U Launcher Lite – FREE Live Cool Themes, Hide Apps', 'ART_AND_DESIGN', '4.7', '87510', '8.7M', '5,000,000+', 'Free', '0', 'Everyone', 'Art & Design', 'August 1, 2018', '1.2.4', '4.0.3 and up']


['Sketch - Draw & Paint', 'ART_AND_DESIGN', '4.5', '215644', '25M', '50,000,000+', 'Free', '0', 'Teen', 'Art & Design', 'June 8, 2018', 'Varies with device', '4.2 and up']


['Pixel Draw - Number Art Coloring Book', 'ART_AND_DESIGN', '4.3', '967', '2.8M', '100,000+', 'Free', '0', 'Everyone', 'Art & Design;Creativity', 'June 20, 2018', '1.1', '4.4 and up']


['Paper flowers instructions', 'ART_AND_DESIGN', '4.4', '167', '5.6M', '50,000+', 'Free', '0', 'Everyone', 'Art & Design', 'March 26, 2017', '1.0', '2.3 and up']


Number of rows:  9614
Number of columns:  13


['284882215', 'Facebook', '389879808', 'USD'

Now, we need to look at only the free apps in these lists as we ony deal with those type of apps. So let's build a way to get rid of the paid apps.

In [17]:
android_final = []
ios_final = []

for app in android_english:
    price = app[7]
    if price == '0':
        android_final.append(app)
        
for app in ios_english:
    price = app[4]
    if price == '0.0':
        ios_final.append(app)
        
explore_data(android_final)
print('\n')
explore_data(ios_final)

['Photo Editor & Candy Camera & Grid & ScrapBook', 'ART_AND_DESIGN', '4.1', '159', '19M', '10,000+', 'Free', '0', 'Everyone', 'Art & Design', 'January 7, 2018', '1.0.0', '4.0.3 and up']


['U Launcher Lite – FREE Live Cool Themes, Hide Apps', 'ART_AND_DESIGN', '4.7', '87510', '8.7M', '5,000,000+', 'Free', '0', 'Everyone', 'Art & Design', 'August 1, 2018', '1.2.4', '4.0.3 and up']


['Sketch - Draw & Paint', 'ART_AND_DESIGN', '4.5', '215644', '25M', '50,000,000+', 'Free', '0', 'Teen', 'Art & Design', 'June 8, 2018', 'Varies with device', '4.2 and up']


['Pixel Draw - Number Art Coloring Book', 'ART_AND_DESIGN', '4.3', '967', '2.8M', '100,000+', 'Free', '0', 'Everyone', 'Art & Design;Creativity', 'June 20, 2018', '1.1', '4.4 and up']


['Paper flowers instructions', 'ART_AND_DESIGN', '4.4', '167', '5.6M', '50,000+', 'Free', '0', 'Everyone', 'Art & Design', 'March 26, 2017', '1.0', '2.3 and up']


Number of rows:  8864
Number of columns:  13


['284882215', 'Facebook', '389879808', 'USD'

Finally, we now have a dataset that we can explore a bit more with. We have 8864 Android apps and 3222 ios apps.

## Most Common Apps

As we are looking for apps that are free, we want to make sure that there is a high level of users to make the app profitable. To accomplish and to minimize risks and overhead, our validation strategy for an app idea is comprised of three steps:

-Build a minimal Android version of the app, and add it to Google Play.
-If the app has a good response from users, we develop it further.
-If the app is profitable after six months, we build an iOS version of the app and add it to the App Store.

Because our end goal is to add the app on both Google Play and the App Store, we need to find app profiles that are successful on both markets. For instance, a profile that works well for both markets might be a productivity app that makes use of gamification.

We want to take a look at the `prime_genre` from the App Store and `Category` and `Genres` from the Google Play Store.

In [24]:
def freq_table(dataset, index):
    table = {}
    
    for app in dataset:
        value = app[index]
        if value in table:
            table[value] += 1
        else:
            table[value] = 1
            
    table_percentage = {}
    for category in table:
        percentage = round((table[category] / len(dataset)) * 100, 2)
        table_percentage[category] = percentage
    
    return table_percentage

def display_table(dataset, index):
    table = freq_table(dataset, index)
    table_display = []
    for key in table:
        key_val_as_tuple = (table[key], key)
        table_display.append(key_val_as_tuple)

    table_sorted = sorted(table_display, reverse = True)
    for entry in table_sorted:
        print(entry[1], ':', entry[0])

So now we have everything set up to get a column, count up each in that column, and sort, we can start look at some datasets. First up is the `prime_genre` category in the App Store

In [25]:
display_table(ios_final, 11)

Games : 58.16
Entertainment : 7.88
Photo & Video : 4.97
Education : 3.66
Social Networking : 3.29
Shopping : 2.61
Utilities : 2.51
Sports : 2.14
Music : 2.05
Health & Fitness : 2.02
Productivity : 1.74
Lifestyle : 1.58
News : 1.33
Travel : 1.24
Finance : 1.12
Weather : 0.87
Food & Drink : 0.81
Reference : 0.56
Business : 0.53
Book : 0.43
Navigation : 0.19
Medical : 0.19
Catalogs : 0.12


As we can see, games clearly are the most popular category among English free apps in the App Store. We can also see that the top three categories are based around entertainment. Let's keep in mind that this is just the amount of apps and does not reflect the most amount of users.

Let's take a look at the Google Play Store by looking at `Category`.

In [26]:
display_table(android_final, 1)

FAMILY : 18.91
GAME : 9.72
TOOLS : 8.46
BUSINESS : 4.59
LIFESTYLE : 3.9
PRODUCTIVITY : 3.89
FINANCE : 3.7
MEDICAL : 3.53
SPORTS : 3.4
PERSONALIZATION : 3.32
COMMUNICATION : 3.24
HEALTH_AND_FITNESS : 3.08
PHOTOGRAPHY : 2.94
NEWS_AND_MAGAZINES : 2.8
SOCIAL : 2.66
TRAVEL_AND_LOCAL : 2.34
SHOPPING : 2.25
BOOKS_AND_REFERENCE : 2.14
DATING : 1.86
VIDEO_PLAYERS : 1.79
MAPS_AND_NAVIGATION : 1.4
FOOD_AND_DRINK : 1.24
EDUCATION : 1.16
ENTERTAINMENT : 0.96
LIBRARIES_AND_DEMO : 0.94
AUTO_AND_VEHICLES : 0.93
HOUSE_AND_HOME : 0.82
WEATHER : 0.8
EVENTS : 0.71
PARENTING : 0.65
ART_AND_DESIGN : 0.64
COMICS : 0.62
BEAUTY : 0.6


We can start to see a much different look on the Google Play Store. There is much more a focus on productivity, even with the large percentage of family apps, which if looked at closer translate to child-based apps.

Let's also look at the `Genres` category.

In [27]:
display_table(android_final, 9)

Tools : 8.45
Entertainment : 6.07
Education : 5.35
Business : 4.59
Productivity : 3.89
Lifestyle : 3.89
Finance : 3.7
Medical : 3.53
Sports : 3.46
Personalization : 3.32
Communication : 3.24
Action : 3.1
Health & Fitness : 3.08
Photography : 2.94
News & Magazines : 2.8
Social : 2.66
Travel & Local : 2.32
Shopping : 2.25
Books & Reference : 2.14
Simulation : 2.04
Dating : 1.86
Arcade : 1.85
Video Players & Editors : 1.77
Casual : 1.76
Maps & Navigation : 1.4
Food & Drink : 1.24
Puzzle : 1.13
Racing : 0.99
Role Playing : 0.94
Libraries & Demo : 0.94
Auto & Vehicles : 0.93
Strategy : 0.91
House & Home : 0.82
Weather : 0.8
Events : 0.71
Adventure : 0.68
Comics : 0.61
Beauty : 0.6
Art & Design : 0.6
Parenting : 0.5
Card : 0.45
Casino : 0.43
Trivia : 0.42
Educational;Education : 0.39
Board : 0.38
Educational : 0.37
Education;Education : 0.34
Word : 0.26
Casual;Pretend Play : 0.24
Music : 0.2
Racing;Action & Adventure : 0.17
Puzzle;Brain Games : 0.17
Entertainment;Music & Video : 0.17
Casual;

The genres category has much more entries, but we can see a very similar landscape. 

What is our main takeaway? In the App Store, free apps are much more geared towards fun, while in the Google Play Store, there is a much more balanced mix of categories.

## Popular Apps by Genre

While it is nice to see what are the categories have the most apps, we really want to see the categories that have the the most users (high popularity).

Google has a catergory for `Installs`, which would show the amount of installs of an app, however, the App Store doesn't have a category like that. There is a category called `rating_count_tot`, which is the total number of user ratings, which might give a clue when used in the case.